src/Target.cpp

```diff
@@ -165,9 +166,47 @@ Target calculate_host_target() {
 #endif
 #endif
 
-    Target t(os, arch, bits, initial_features);
+    return {os, arch, bits, initial_features};
+}
+
+int get_cuda_capability_lower_bound(const Target &t) {
+    if (!t.has_feature(Target::CUDA)) {
+        return 0;
+    }
+    if (t.has_feature(Target::CUDACapability30)) {
+        return 30;
+    }
+    if (t.has_feature(Target::CUDACapability32)) {
+        return 32;
+    }
+    if (t.has_feature(Target::CUDACapability35)) {
+        return 35;
+    }
+    if (t.has_feature(Target::CUDACapability50)) {
+        return 50;
+    }
+    if (t.has_feature(Target::CUDACapability61)) {
+        return 61;
+    }
+    return 20;
+}
 
-    return Target(os, arch, bits, initial_features);
+int get_hvx_lower_bound(const Target &t) {
+    if (!t.has_feature(Target::HexagonDma) && t.arch != Target::Hexagon) {
```

Comment:
You shouldn't  need to check ``Target::HexagonDma`` in relation with hvx_lower_bound. For compiling a runtime on hexagon ``t.arch`` will _never_ perhaps be anything other than ``Target::Hexagon`` and so, you'll not likely satisfy this condition. I think it is simply safe to leave ``Target::HexagonDma`` in the union list.

Reply:
It would be really useful if someone could explain the various Hexagon flags / arch setting to me. I have never worked with Hexagon, ever, so I'm well aware there are going to be issues with the implementation until someone helps me fix it.

Reply:
Sure. 

- HVX_64 and HVX_128 allow you to choose between 64 byte vectors and 128 byte vectors. Only **one of these** should be present
- HVX_v62, HVX_v65 and HVX_v65 are successive ISA generations, each a super set of generations before it. So, a tighter (lower) bound is the right thing to do.
- HVX_shared_object - I think it is best to remove this altogether. I don't believe it is used any more.
- HexagonDma - This includes runtime functions to support Direct Memory Access functionality and as such is decoupled from the choice of HVX_v**.

Reply:
Thanks, @pranavb-ca -- how are these flags affected by the presence / absence of hexagon as the arch? Are these flags even valid if `arch == Linux`, say?

Reply:
Also, is there a generation _before_ HVX_v62? Is it valid to have none of those three flags set? If so, what does it mean?

Reply:
Typical target values we use most often for compiling Hexagon runtimes are
``hexagon-32-qurt-hvx_128`` or ``hexagon-32-qurt-hvx_128-hexagon_dma`` (or the same ones with ``hvx_64`` in place of ``hvx_128``

Reply:
> Also, is there a generation _before_ HVX_v62, analogous to CUDA Capability 20? Is it valid to have none of those three flags set? If so, what does it mean?

The Hexagon backend in Halide defaults to v60. It isn't a separate target feature though. So, haveing v62,65, 66 reset means you get v60.

Reply:
Thanks so much again! Very helpful. Could you answer my questions about other architectures? Is it valid to have `x86-64-linux-hvx_128-hexagon_dma`?

Reply:
> Thanks, @pranavb-ca -- how are these flags affected by the presence / absence of hexagon as the arch? Are these flags even valid if `arch == Linux`, say?

``os == Linux`` when ``arch == Hexagon`` will lead to problems because the OS on Hexagon is ``QuRT``. So, it doesn't have posix like functionality. However, you won't be changing any behaviour by not handling this bad combination.

Typically Generators can be compiled with ``Arch != Target::Hexagon``, for eg. ``arm-64-android-hvx_128``. This simply means that the pipeline will be offloaded to the Hexagon however, even in these cases the runtime is compiled using ``hexagon-32-qurt-hvx_128`` (See HexagonOffload.cpp)

Reply:
Okay, it looks like the compilation process will generate a runtime if necessary using the features on the target. So we can merge `arm-64-android` and `arm-64-android-hvx_128` to get the latter? And `arm-64-android` does not imply hexagon v60, it's just no hexagon in the absence of any other hexagon flag?

Reply:
Oh wait, any use of Hexagon requires exactly one of `HVX_64` or `HVX_128` to be set. I can just check that, right?

Reply:
> Oh wait, any use of Hexagon requires exactly one of `HVX_64` or `HVX_128` to be set. I can just check that, right?

Yes

Reply:
Awesome -- could you take another look at the code? I think it matches the description you gave.

Reply:
> > Oh wait, any use of Hexagon requires exactly one of `HVX_64` or `HVX_128` to be set. I can just check that, right?

> Yes

Are you sure about this? It might not be a typical usage of Halide, but I think it should be totally possible to build code with arch == Hexagon and no HVX flags.

Reply:
@dsharletg - what happens in that case?

Reply:
> > > Oh wait, any use of Hexagon requires exactly one of `HVX_64` or `HVX_128` to be set. I can just check that, right?
> 
> > Yes
> 
> Are you sure about this? It might not be a typical usage of Halide, but I think it should be totally possible to build code with arch == Hexagon and no HVX flags.

Sorry, I misread/misunderstood Alex's comment- I was talking about how they should be mutually exclusive. Neither is necessary. ``arch == Target::Hexagon`` implies v60. Although not intended, but Alex's code still ensures this as the absence of HVX_64 and HVX_128 results in teh resetting of 62,65 and 66. Leaving a default of v60 that is ensured by CodeGen_Hexagon.cpp.

However, to be strictly correct we should also check if ``arch == Hexagon``

Reply:
aargh. @alexreinking  - I am really sorry for making you go over this again :(

@dsharletg  is absolutely right. To `return 0` from ``get_hvx_lower_bound`` your check needs to be
``if (!t.has_feature(Target::HVX_64) && !t.has_feature(Target::HVX_128) && t.arch != Target::Hexagon)``

Reply:
@dsharletg, @pranavb-ca -- Is this target valid: `arm-32-android-hvx_v62`? If so, is `hvx_64` or `hvx_128` set? What about when we have `hexagon-32-qurt-hvx_v62`?

Reply:
> @dsharletg - what happens in that case?

So, v60, v61 etc are Scalar ISA versions. We don't have these features in Halide because Halide for the Scalar ISA is not interesting. hvx_v6x is Scalar ISA + Hexagon Vector Extensions.

So,  ``hvx_64`` and ``hvx_128`` are useful really only if you use ``.vectorize()``
And a target like ``hexagon-32-qurt-hvx_v62` will result in v62 scalar code.

Reply:
> @dsharletg, @pranavb-ca -- Is this target valid: `arm-32-android-hvx_v62`? If so, is `hvx_64` or `hvx_128` set? What about when we have `hexagon-32-qurt-hvx_v62`?

For ``arm-32-android-hvx_v62`` - This is valid and should result in v62 . If the hexagon target is instantiated, which in this case will happen, then hvx_v64 is the default. I think we should change this default, but that is outside the scope of this discussion. So effectively, you get ``hvx_64-hvx_v62``

``hexagon-32-qurt-hvx_v62`` - Again, hvx_v64 default and v62 ISA.

Reply:
So the check should _really_ look for _any_ of the `hvx_*` flags or `arch == Hexagon`

Reply:
Actually, does `arm-32-android-hexagon_dma` equal `arm-32-android-hvx_64-hexagon_dma` with HVX_v60?

Reply:
Should it really be:

```
bool is_using_hexagon(const Target &t) {
    return t.has_feature(Target::HVX_64)
           || t.has_feature(Target::HVX_128)
           || t.has_feature(Target::HVX_v62)
           || t.has_feature(Target::HVX_v65)
           || t.has_feature(Target::HVX_v66)
           || t.has_feature(Target::HexagonDma)
           || t.has_feature(Target::HVX_shared_object)
           || t.arch == Target::Hexagon;
}

int get_hvx_lower_bound(const Target &t) {
    if (!is_using_hexagon(t)) {
        return 0;
    }
    if (t.has_feature(Target::HVX_v62)) {
        return 62;
    }
    if (t.has_feature(Target::HVX_v65)) {
        return 65;
    }
    if (t.has_feature(Target::HVX_v66)) {
        return 66;
    }
    return 60;
}
```

Reply:
> Should it really be:
> 
> ```
> bool is_using_hexagon(const Target &t) {
>     return t.has_feature(Target::HVX_64)
>            || t.has_feature(Target::HVX_128)
>            || t.has_feature(Target::HVX_v62)
>            || t.has_feature(Target::HVX_v65)
>            || t.has_feature(Target::HVX_v66)
>            || t.has_feature(Target::HexagonDma)
>            || t.has_feature(Target::HVX_shared_object)
>            || t.arch == Target::Hexagon;
> }
> 
> int get_hvx_lower_bound(const Target &t) {
>     if (!is_using_hexagon(t)) {
>         return 0;
>     }
>     if (t.has_feature(Target::HVX_v62)) {
>         return 62;
>     }
>     if (t.has_feature(Target::HVX_v65)) {
>         return 65;
>     }
>     if (t.has_feature(Target::HVX_v66)) {
>         return 66;
>     }
>     return 60;
> }
> ```

Yes (I should submit a PR to remove HVX_shared_object)

Reply:
> Yes

Excellent. I just pushed

> (I should submit a PR to remove HVX_shared_object)

I look forward to it!